client/client.go

```diff
@@ -99,56 +123,513 @@ func UploadAdd(issuer principal.Signer, space did.DID, params uploadadd.Caveat,
+	}
+
+	// TODO: define a schema and use bindnode.Rebind rather than doing this manually
+	var id did.DID
+	keys := map[string][]byte{}
+	it := putKeysNode.MapIterator()
+	for !it.Done() {
+		k, v, err := it.Next()
+		if err != nil {
+			return fmt.Errorf("invalid put facts: %w", err)
+		}
+
+		kStr, err := k.AsString()
+		if err != nil {
+			return fmt.Errorf("invalid put facts: %w", err)
+		}
+
+		switch kStr {
+		case "id":
+			// v is a did string
+			vStr, err := v.AsString()
+			if err != nil {
+				return fmt.Errorf("invalid put facts: %w", err)
+			}
+
+			id, err = did.Parse(vStr)
+			if err != nil {
+				return fmt.Errorf("invalid put facts: %w", err)
+			}
+		case "keys":
+			// v is a did to key map
+			it2 := v.MapIterator()
+			for !it2.Done() {
+				k2, v2, err := it2.Next()
+				if err != nil {
+					return fmt.Errorf("invalid put facts: %w", err)
+				}
+
+				k2Str, err := k2.AsString()
+				if err != nil {
+					return fmt.Errorf("invalid put facts: %w", err)
+				}
+
+				v2Bytes, err := v2.AsBytes()
+				if err != nil {
+					return fmt.Errorf("invalid put facts: %w", err)
+				}
+
+				keys[k2Str] = v2Bytes
+			}
+		}
+	}
+
+	derivedKey, ok := keys[id.String()]
+	if !ok {
+		return fmt.Errorf("invalid put facts: missing key for %s", id.String())
+	}
+
+	derivedSigner, err := signer.Decode(derivedKey)
+	if err != nil {
+		return fmt.Errorf("deriving signer: %w", err)
+	}
+
+	putRcpt, err := receipt.Issue(derivedSigner, result.Ok[httpcap.PutOk, ipld.Builder](httpcap.PutOk{}), ran.FromInvocation(putTask))
+	if err != nil {
+		return fmt.Errorf("generating receipt: %w", err)
+	}
+
+	// var concludeFacts []ucan.FactBuilder
+	// for rcptBlock, err := range putRcpt.Blocks() {
+	// 	if err != nil {
+	// 		return nil, nil, fmt.Errorf("getting receipt block: %w", err)
+	// 	}
+
+	// 	concludeFacts = append(concludeFacts, rcptBlock.Link())
+	// }
+
+	httpPutConcludeInvocation, err := ucancap.Conclude.Invoke(
+		issuer,
+		audience,
+		issuer.DID().String(),
+		ucancap.ConcludeCaveats{
+			Receipt: putRcpt.Root().Link(),
+		},
+		// delegation.WithFacts(concludeFacts),
+		delegation.WithNoExpiration(),
+	)
+	if err != nil {
+		return fmt.Errorf("generating invocation: %w", err)
+	}
+
+	// attach the receipt to the conclude invocation
+	for rcptBlock, err := range putRcpt.Blocks() {
+		if err != nil {
+			return fmt.Errorf("getting receipt block: %w", err)
+		}
+
+		httpPutConcludeInvocation.Attach(rcptBlock)
+	}
+
+	resp, err := client.Execute([]invocation.Invocation{httpPutConcludeInvocation}, conn)
+	if err != nil {
+		return fmt.Errorf("executing conclude invocation: %w", err)
+	}
+
+	rcptlnk, ok := resp.Get(httpPutConcludeInvocation.Link())
+	if !ok {
+		return fmt.Errorf("receipt not found: %s", httpPutConcludeInvocation.Link())
+	}
+
+	reader, err := receipt.NewReceiptReaderFromTypes[ucancap.ConcludeOk, fdm.FailureModel](ucancap.ConcludeOkType(), fdm.FailureType(), captypes.Converters...)
```

Comment:
I might consider putting stuff like this outside the functions, a.l.a.

```golang
var receiptReader ReceiptReader[ucancap.ConcludeOk, fdm.FailureModel]

func init() {
   receiptReader, err := receipt.NewReceiptReaderFromTypes[ucancap.ConcludeOk, fdm.FailureModel](ucancap.ConcludeOkType(), fdm.FailureType(), captypes.Converters...)
   if err != nil {
     panic(err)
   }
}
```

In fact, I might go a step futher and just put these in go-libstoracha to avoid having to make them over and over, just like we do with the capabilities

Reply:
I like the idea. The original capabilities implemented in `go-w3up` did this.

Captured in https://github.com/storacha/go-libstoracha/issues/20